### PR TITLE
HaplotypeCaller: tests for allele specific annotations

### PR DESCRIPTION
**Pull request by akiezun**
this PR matches the branch ak_original_allelespecificannotations in https://github.com/broadinstitute/gatk/pull/1840

@droazen can you review? It matches the 3.5 very closely. There are some off by one diffs in DP and PLs that are probably trivial and inconsequential. AS_ annotations in VCF mode produces results very different from 3.5 (we make raw annotations and 3.5 combines them) but it's not a big deal for now.

### CODE REVIEW

**Comment by akiezun** [comment]
broadinstitute/gatk#1840 is now merged so protected should be reved up and this branch should compile then

**Comment by akiezun** [comment]
@droazen should be ready now

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 1]
```diff
@@ -43,15 +43,50 @@ public void testVCFModeIsConsistentWithPastResults() throws Exception {
...
     /*
-     * Test that in VCF mode we're >= 99% concordant with GATK3.4 results
+     * Test that in VCF mode we're consistent with past GATK4 results
      */
     @Test
-    public void testVCFModeIsConcordantWithGATK3Results() throws Exception {
+    public void testVCFModeIsConsistentWithPastResults_AlleleSpecificAnnotations() throws Exception {
+        Utils.resetRandomGenerator();
+
+        //NOTE: AlleleSpecific support in the VCF mode is implemented but bous for now.
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: `bous` -> `bogus`. Also, if the AS output in VCF mode is bogus, we should probably be throwing a `UserException` for now in `HaplotypeCaller` when the user requests that combination, and open a ticket to address.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 2]
```diff
@@ -64,7 +99,40 @@ public void testVCFModeIsConcordantWithGATK3Results() throws Exception {
         runCommandLine(args);
 
         final double concordance = calculateConcordance(output, gatk3Output);
-        Assert.assertTrue(concordance >= 0.99, "Concordance with GATK 3.4 in VCF mode is < 99%");
+        Assert.assertTrue(concordance >= 0.99, "Concordance with GATK 3.5 in VCF mode is < 99%");
+    }
+
+    /*
+ * Test that in VCF mode we're >= 99% concordant with GATK3.5 results
+ */
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: Fix indentation

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 3]
```diff
@@ -43,15 +43,50 @@ public void testVCFModeIsConsistentWithPastResults() throws Exception {
...
+
+    /*
+     * Test that in VCF mode we're >= 99% concordant with GATK3.5 results
+     */
+    @Test
+    public void testVCFModeIsConcordantWithGATK3_5Results() throws Exception {
         Utils.resetRandomGenerator();
 
         final File output = createTempFile("testVCFModeIsConcordantWithGATK3Results", ".vcf");
-        final File gatk3Output = new File(TEST_FILES_DIR, "expected.testVCFMode.gatk3.4.vcf");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: You left the previous expected output `expected.testVCFMode.gatk3.4.vcf` checked into the test directory -- should we delete it, or is it worth keeping around for a while for troubleshooting purposes?

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 4]
```diff
@@ -93,36 +161,99 @@ public void testGVCFModeIsConsistentWithPastResults() throws Exception {
     }
 
     /*
+ * Test that in GVCF mode we're consistent with past GATK4 results using AS_ annotations
+ */
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: Fix indentation (here and elsewhere)

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 5]
```diff
@@ -93,36 +161,99 @@ public void testGVCFModeIsConsistentWithPastResults() throws Exception {
...
+    /*
      * Test that in GVCF mode we're >= 99% concordant with GATK3 results
      */
     @Test
-    public void testGVCFModeIsConcordantWithGATK3Results() throws Exception {
+    public void testGVCFModeIsConcordantWithGATK3_5Results() throws Exception {
         Utils.resetRandomGenerator();
 
         final File output = createTempFile("testGVCFModeIsConcordantWithGATK3Results", ".g.vcf");
-        final File gatk3Output = new File(TEST_FILES_DIR, "expected.testGVCFMode.gatk3.4.g.vcf");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: You left the previous expected output `expected.testGVCFMode.gatk3.4.g.vcf` checked into the test directory -- should we delete it, or is it worth keeping around for a while for troubleshooting purposes?

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 6]
```diff
@@ -64,7 +99,40 @@ public void testVCFModeIsConcordantWithGATK3Results() throws Exception {
...
+                "-O", output.getAbsolutePath(),
+                "-G", "StandardAnnotation",
+                "-G", "AS_StandardAnnotation",
+                "-pairHMM", "AVX_LOGLESS_CACHING"
+        };
+
+        runCommandLine(args);
+
+        final double concordance = calculateConcordance(output, gatk3Output);
+        Assert.assertTrue(concordance >= 0.99, "Concordance with GATK 3.5 in VCF mode is < 99%");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: `in VCF mode` -> `in AS VCF mode`

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 7]
```diff
@@ -93,36 +161,99 @@ public void testGVCFModeIsConsistentWithPastResults() throws Exception {
...
+        Utils.resetRandomGenerator();
+        final File output = createTempFile("testGVCFModeIsConcordantWithGATK3_5AlelleSpecificResults", ".g.vcf");
+        final File gatk3Output = new File(TEST_FILES_DIR + "expected.testGVCFMode.gatk3.5.alleleSpecific.g.vcf");
+
+        //Created by running:
+        // java -jar ~/bin/GenomeAnalysisTK-3.5.0/GenomeAnalysisTK.jar -T HaplotypeCaller \
+        // -I ./src/test/resources/large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam \
+        // -R src/test/resources/large/human_g1k_v37.20.21.fasta -L 20:10000000-10100000 \
+        // -ERC GVCF --out as.gatk3.5.noDownsample.g.vcf -G StandardHC -G Standard -G AS_Standard \
+        // --disableDithering --no_cmdline_in_header  -dt NONE --maxReadsInRegionPerSample 100000000 --minReadsPerAlignmentStart 100000
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: Comment should be before the `gatk3Output` var

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 8]
```diff
@@ -93,36 +161,99 @@ public void testGVCFModeIsConsistentWithPastResults() throws Exception {
...
+        Assert.assertTrue(concordance >= 0.99, "Concordance with GATK 3.5 in AS GVCF mode is < 99%");
     }
 
     @Test
     public void testBamoutProducesReasonablySizedOutput() {
         Utils.resetRandomGenerator();
 
         // We will test that when running with -bamout over the testInterval, we produce
-        // a bam with a number of reads that is within 10% of what GATK3.4 produces with
+        // a bam with a number of reads that is within 10% of what GATK3.5 produces with
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: Did you update `gatk3BamoutNumReads` in this test to reflect the number of reads that GATK3.5 produces with `-bamout`?

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java [hunk 9]
```diff
@@ -64,7 +99,40 @@ public void testVCFModeIsConcordantWithGATK3Results() throws Exception {
...
+    public void testVCFModeIsConcordantWithGATK3_5ResultsAlleleSpecificAnnotations() throws Exception {
+        Utils.resetRandomGenerator();
+
+        final File output = createTempFile("testVCFModeIsConcordantWithGATK3_5ResultsAlleleSpecificAnnotations", ".vcf");
+
+        //Created by running
+        //java -jar ~/bin/GenomeAnalysisTK-3.5.0/GenomeAnalysisTK.jar -T HaplotypeCaller \
+        // -I ./src/test/resources/large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam \
+        // -R src/test/resources/large/human_g1k_v37.20.21.fasta -L 20:10000000-10100000 \
+        // --out as.gatk3.5.noDownsample.vcf -G StandardHC -G Standard -G AS_Standard \
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: Can you explain why you run with `-G StandardHC` in GATK3 but not GATK4?

**Comment by droazen** [comment]
Review complete, minor comments only. Back to @akiezun, merge after addressing comments.

**Comment by lbergelson** [comment]
@akiezun Genotypegvcfs is blocked on this since it rebased.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java (hunk 1 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: ok

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java (hunk 2 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java (hunk 3 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: I think we should delete now that you've reviewed it

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java (hunk 4 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: fixed

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java (hunk 5 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: I think we should delete now that you've reviewed it

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java (hunk 6 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: fixed

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java (hunk 7 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: fixed

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java (hunk 8 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: I just checked - 3.5 also produces 5170 reads

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java (hunk 9 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: yes, the difference stems from the fact that gatk4 interprets collection arguments as 'add to the list of defaults' while gatk3 interpreted them as 'replace the list of defaults with the provided list'. We may (probably should) change that but that's a different ticket.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java (hunk 1 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: https://github.com/broadinstitute/gatk-protected/issues/576

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java (hunk 9 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerIntegrationTest.java]: here's the new ticket https://github.com/broadinstitute/gatk/issues/1928

**Pull request merged**